### PR TITLE
Add signs for Lsp diagnostics

### PR DESCRIPTION
**Pull request by pwntester**

### CODE REVIEW

File: runtime/doc/lsp.txt [hunk 1]
```diff
@@ -878,6 +878,18 @@ buf_diagnostics_underline({bufnr}, {diagnostics})
 buf_diagnostics_virtual_text({bufnr}, {diagnostics})
                 TODO: Documentation
 
+                                 *vim.lsp.util.buf_diagnostics_signs()*
+buf_diagnostics_signs({bufnr}, {diagnostics})
+                Place signs for each diagnostic in the sign column.
+                Sign characters can be customized with the following options:
+>
+let g:LspDiagnosticsErrorSign = 'E'
+let g:spDiagnosticsWarningSign = 'W'
```
Comment (psafont) [inline runtime/doc/lsp.txt]: Typo: should be ` g:LspDiagnosticsWarningSign`

Comment (pwntester) [reply runtime/doc/lsp.txt]: Thanks!

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -688,6 +699,19 @@ do
...
+    for _, diagnostic in ipairs(diagnostics) do
+      if diagnostic.severity == protocol.DiagnosticSeverity.Error then
+        vim.fn.sign_place(0, sign_ns, 'LspDiagnosticsErrorSign', bufnr, {lnum=(diagnostic.range.start.line+1)})
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Warning then
+        vim.fn.sign_place(0, sign_ns, 'LspDiagnosticsWarningSign', bufnr, {lnum=(diagnostic.range.start.line+1)})
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Information then
+        vim.fn.sign_place(0, sign_ns, 'LspDiagnosticsInformationSign', bufnr, {lnum=(diagnostic.range.start.line+1)})
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Hint then
+        vim.fn.sign_place(0, sign_ns, 'LspDiagnosticsHintSign', bufnr, {lnum=(diagnostic.range.start.line+1)})
+      end
```
Comment (teto) [inline runtime/lua/vim/lsp/util.lua]: the call remains the same apart from the hl name. Either assign the name into avariable or create a function to map the level to the highlight name or just change the if.
`vim.fn.sign_place(0, sign_ns, <NAME>, bufnr, {lnum=(diagnostic.range.start.line+1)})` should appear once.

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -545,6 +545,12 @@ do
   local all_buffer_diagnostics = {}
 
   local diagnostic_ns = api.nvim_create_namespace("vim_lsp_diagnostics")
+  local sign_ns = 'vim_lsp_signs'
+
+  vim.fn.sign_define('LspDiagnosticsErrorSign',       {text=vim.g['LspDiagnosticsErrorSign'] or 'x', texthl='LspDiagnosticsError', linehl='', numhl=''})
```
Comment (teto) [inline runtime/lua/vim/lsp/util.lua]: this could use better defaults than just "x", like the same you used in the help.

**Comment by pwntester** [comment]
Thanks for the feedback. PR rebased and comments addressed, should be ready for merging?

**Comment by teto** [comment]
I've tried overriding `let g:LspDiagnosticsErrorSign = 'T' in my init.vim` but this keeps using `E` as a sign, I wonder if it's because it is defined too late since `lua print(vim.g['LspDiagnosticsErrorSign'])` does print "T".

**Comment by wsdjeg** [comment]
> let g:LspDiagnosticsErrorSign = 'E'
let g:LspDiagnosticsWarningSign = 'W'
let g:LspDiagnosticsInformationSign = 'I'
let g:LspDiagnosticsHintSign = 'H'

How about use lua option instead of viml option? `g:LspDiagnosticsHintSign` this maybe used by other plugin.

**Comment by pwntester** [comment]
> I've tried overriding let g:LspDiagnosticsErrorSign = 'T' in my init.vim but this keeps using E as a sign, I wonder if it's because it is defined too late since lua print(vim.g['LspDiagnosticsErrorSign']) does print "T".

@teto @justinmk  any ideas what is the best way to fix this? should I redifne the signs everytime diagnostics are rendered?

**Comment by teto** [comment]
Just defer the call so that init.vim is loaded. In buf_diagnostics_signs for instance. or on `on_attach`.

**Comment by teto** [comment]
please rebase

**Comment by pwntester** [comment]
Seems like build error is a timeout rather than an error

**Comment by teto** [comment]
tested locally with success, will merge once CI finishes.

**Comment by teto** [comment]
thanks

**Comment by Kyle-Thompson** [comment]
Is there an option to disable signs? I would personally like to avoid anything that creates a new column and shifts all my code over.

**Comment by teto** [comment]
I am not sure what interface we will favor to users or if that will be left to yet another plugin but for now you can override the callbacks so that it doesnt run 
https://github.com/neovim/neovim/blob/49cd750d6a72efc0571a89d7a874bbb01081227f/runtime/lua/vim/lsp/callbacks.lua#L35

For instance like this https://github.com/teto/home/blob/fae521dac3a131c529984da9f53558f892c5cbb3/config/nvim/init.lua#L154 (copy/paste the original callback without the call `util.buf_diagnostics_signs(bufnr, result.diagnostics)`)
Not yet for now you can customize your callbacks so that it doesn't run the default one

**Comment by Kyle-Thompson** [comment]
I pasted the contents of the do block from your init.lua within a lua heredoc directly under my nvim_lsp config and I'm still getting the sign.

```
lua << EOF
local nvim_lsp = require('nvim_lsp')
nvim_lsp.clangd.setup{
  cmd = {'clangd', '--background-index',
         '--clang-tidy', '--log=error', '--pretty'};
}

do
  local method = 'textDocument/publishDiagnostics'
  local default_callback = vim.lsp.callbacks[method]
vim.lsp.callbacks[method] = function(err, method, result, client_id)
    default_callback(err, method, result, client_id)
    if result and result.diagnostics then
      for _, v in ipairs(result.diagnostics) do
        v.uri = v.uri or result.uri
      end
      vim.lsp.util.set_loclist(result.diagnostics)
    end
  end
end
EOF
```
How else should I be doing this?

**Comment by teto** [comment]
remove the call to the default callback
`default_callback(err, method, result, client_id)`

**Comment by Kyle-Thompson** [comment]
Removing that removes all diagnostics. I still want the virtual text, just no signs.

**Comment by teto** [comment]
See the default callback I had linked in https://github.com/neovim/neovim/pull/11668#issuecomment-592974004, you can copy/paste the actions you want, e.g., for diagnostics:
`util.buf_diagnostics_virtual_text(bufnr, result.diagnostics)`

**Comment by Kyle-Thompson** [comment]
Thanks!
To anyone like me who's pretty unfamiliar with lua and wants something easy to copy paste, this worked for me.

```
lua << EOF
local nvim_lsp = require('nvim_lsp')
local util = require 'vim.lsp.util'
nvim_lsp.clangd.setup{}

do
  local method = 'textDocument/publishDiagnostics'
  vim.lsp.callbacks[method] = function(_, _, result)
    if not result then return end
    local uri = result.uri
    local bufnr = vim.uri_to_bufnr(uri)
    if not bufnr then
      err_message("LSP.publishDiagnostics: Couldn't find buffer for ", uri)
      return
    end
    util.buf_clear_diagnostics(bufnr)
    util.buf_diagnostics_save_positions(bufnr, result.diagnostics)
    util.buf_diagnostics_underline(bufnr, result.diagnostics)
    util.buf_diagnostics_virtual_text(bufnr, result.diagnostics)
    vim.api.nvim_command("doautocmd User LspDiagnosticsChanged")
  end
end
EOF
```

**Comment by teto** [comment]
@Kyle-Thompson this was advertised on reddit recently and I thought it might interest you:
https://github.com/haorenW1025/diagnostic-nvim

**Pull request merged**